src/memory_coloring.cpp

```diff
@@ -0,0 +1,381 @@
+                       [](auto&& pp) { return pp.first; });
+
+        // Sort the conflict queue so we process the allocation with the least
+        // number of adjacent allocations first
+        std::sort(conflict_queue.begin(), conflict_queue.end(), [&](auto x, auto y) {
+            return std::make_tuple(conflict_table.at(x).size(), x->get_shape().bytes()) <
+                   std::make_tuple(conflict_table.at(y).size(), y->get_shape().bytes());
+        });
+        // Process the conflict_queue, we refer to the current allocation as
+        // the parent and the adjacent allocations as children
+        for(auto parent : conflict_queue)
+        {
+            // Sort children by size
+            std::vector<instruction_ref> children(conflict_table.at(parent).begin(),
+                                                  conflict_table.at(parent).end());
+            std::sort(children.begin(), children.end(), [](auto x, auto y) {
+                return x->get_shape().bytes() < y->get_shape().bytes();
+            });
+            assert(not contains(children, parent));
+            // This set is to track the segments already processed
+            std::set<segment> segments;
+            // Add all segments for the children to the segments already processed
+            transform_if(
+                children.begin(),
+                children.end(),
+                std::inserter(segments, segments.begin()),
+                [&](auto child) { return as.get_segment(child); },
+                [&](auto child) { return *as.get_segment(child); });
+
+            // Get the segment for the parent
+            const auto* parent_segment = as.get_segment(parent);
+            // Add segment for the parent if there is none or segment overlaps with the children
+            if(parent_segment == nullptr or overlaps(segments, *parent_segment))
+                as.add_segment(parent, next_segment(segments, parent, alignment));
+        }
+        // Reduce the number of segments
+        for(std::size_t n = 0; n < 3; n++)
+        {
+            for(auto parent : conflict_queue)
+            {
+                auto children = conflict_table.at(parent);
+                // This set is to track the segments already processed
+                std::set<segment> segments;
+                // Add all segments for the children to the segments already processed
+                transform_if(
+                    children.begin(),
+                    children.end(),
+                    std::inserter(segments, segments.begin()),
+                    [&](auto child) { return as.get_segment(child); },
+                    [&](auto child) { return *as.get_segment(child); });
+                // Get the segment for the parent
+                const auto* parent_segment = as.get_segment(parent);
+                assert(parent_segment != nullptr);
+
+                auto s = next_segment(segments, parent, alignment);
```

Comment:
The alignment has to be the same for all segments.

Reply:
Yes, I understand that, but I think this pass trying to do two things at once, which are alignment and memory allocation, which may not result in optimal allocation in terms of fragmentation. I was  thinking if it would be possible to just do (1-byte) alignment in this pass and adjust those alignments later in another pass separating two concerns.

Reply:
>  but I think this pass trying to do two things at once, which are alignment and memory allocation

I dont see how you can seperate these two.

> which may not result in optimal allocation in terms of fragmentation

What do you mean? The only way to possibly get better memory for alignment is to use a different alignment for every segment. Since most models all use the same data type, I dont see the reason to introduce this complexity.

> I was thinking if it would be possible to just do (1-byte) alignment in this pass and adjust those alignments later in another pass separating two concerns.

This definitely wont work. If we change alignment then we could create overlaps of the segments since we need to shift the segments forward or backwards to be aligned. 

Also, segments are part of the internal data structures of memory coloring so I dont see how a later pass could adjust that, since it wont have access to it.